### PR TITLE
CovarianceModel updating

### PR DESCRIPTION
**Pull request by sofianehaddad**
The purpose of this pull-request is to improve robustness, performance of covariance models & make benefit inside `Kriging`, `KL` and `GP` classes

Several tasks have been achieved:
 - Removed `computeStandardRepresentative` which was badly used inside `1d/nD` models. Now scalar models rely on `computeAsScalar` and nD models on `operator()`. 
 - 'Hybrid' models (models that might be `1d`/`nD` or `stationary/non stationary`) should implement **all** required methods
 - Removed useless `StationaryCovarianceModel` : it is a behaviour and should not be used as inheritance
 - TensorizedCovarianceModel now relies only on a collection of 1d output models. Future improvement will be to introduce/account a correlation
 - Fix several bugfixes/issues

### CODE REVIEW

File: lib/src/Base/Stat/CovarianceModelImplementation.cxx [hunk 1]
```diff
@@ -118,6 +121,7 @@ CovarianceModelImplementation::CovarianceModelImplementation(const Point & scale
   , outputCovarianceCholeskyFactor_(0)
   , definesComputeStandardRepresentative_(false)
   , isDiagonal_(true)
+  , isStationary_(false)
```
Comment (jschueller) [inline lib/src/Base/Stat/CovarianceModelImplementation.cxx:120]: now with c++11 you can write "Bool isStationary_ = false" in the .hxx and you're done

**Review by jschueller — APPROVED** [review]

File: lib/src/Base/Stat/CovarianceModelImplementation.cxx (hunk 1 above)
Comment (sofianehaddad) [reply lib/src/Base/Stat/CovarianceModelImplementation.cxx:120]: yep thanks. But the idea is to have similarities with other attributes

**Comment by regislebrun** [comment]
Hi @sofianehaddad ,

I tested your branch against master on the following benchmark:
```
from openturns import *
from time import time

def decomposeCube(xMin, xMax, BIG = 10000):
    print("Discretization benchmark, BIG=", BIG)
    dx = xMax[0] - xMin[0]
    dy = xMax[1] - xMin[1]
    dz = xMax[2] - xMin[1]
    N = (BIG/(dx*dy*dz))**(1.0/3.0)
    Nx = int(N*dx)
    Ny = int(N*dy)
    Nz = int(N*dz)
    print("Nx=", Nx, "Ny=", Ny, "Nz=", Nz)
    cube = IntervalMesher([Nx, Ny, Nz]).build(Interval(xMin, xMax))
    cube3D = IntervalMesher([Nx//2, Ny//2, Nz//2]).build(Interval(xMin, xMax))
    print("Cube vertices=", cube.getVerticesNumber())
    param = HMatrixParameters()
    covariance = SquaredExponential([0.1 * dx, 0.1 * dy, 0.1 * dz], [0.02*(dx+dy+dz)])
    print("#"*50)
    print("Scalar case, dense")
    t0 = time()
    res = covariance.discretize(cube)
    t1 = time()
    print("t=", t1 - t0, "s")
    print("#"*50)
    print("Scalar case, hmat")
    t0 = time()
    res = covariance.discretizeHMatrix(cube, param)
    t1 = time()
    print("t=", t1 - t0, "s")
    covariance = TensorizedCovarianceModel([SquaredExponential([0.1 * dx, 0.1 * dy, 0.1 * dz], [0.02*(dx+dy+dz)])]*3)
    print("#"*50)
    print("Multivariate case, dense")
    t0 = time()
    res = covariance.discretize(cube3D)
    t1 = time()
    print("t=", t1 - t0, "s")
    print("#"*50)
    print("Multivariate case, hmat")
    t0 = time()
    res = covariance.discretizeHMatrix(cube3D, param)
    t1 = time()
    print("t=", t1 - t0, "s")

decomposeCube([0.0]*3, [1.0]*3, BIG=15000)
```
and the results are here for master:
```
Discretization benchmark, BIG= 15000
Nx= 24 Ny= 24 Nz= 24
Cube vertices= 15625
##################################################
Scalar case, dense
t= 5.715564012527466 s
##################################################
Scalar case, hmat
t= 1.827186107635498 s
##################################################
Multivariate case, dense
t= 0.4884326457977295 s
##################################################
Multivariate case, hmat
t= 24.384877681732178 s
```
and for your branch:
```
Discretization benchmark, BIG= 15000
Nx= 24 Ny= 24 Nz= 24
Cube vertices= 15625
##################################################
Scalar case, dense
t= 3.102839231491089 s
##################################################
Scalar case, hmat
t= 1.651404619216919 s
##################################################
Multivariate case, dense
t= 0.4152414798736572 s
##################################################
Multivariate case, hmat
t= 21.710374355316162 s
```
The improvement is very significant for the 1D/Dense case, and good to take for the other cases. The main bottleneck I see is the trampoline_compute method, which is not aware of the fact that the output covariance matrix for a tensorized model is diagonal.

**Review by regislebrun — APPROVED** [review]
Great job!

**Comment by sofianehaddad** [comment]
@regislebrun Indeed. Let us keep it in mind as further improvement! I have some ones in mind that should be discussed (for 1.18 probably)

**Pull request merged**